crypto/ml_kem/ml_kem-c.inc

```diff
@@ -0,0 +1,240 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright (c) 2024, Google Inc. */
+
+#include <stdlib.h>
+#include <openssl/e_os2.h>
+#include <openssl/rand.h>
+#include <internal/sha3.h>
+#include <internal/constant_time.h>
+#include <internal/common.h>
+#include <crypto/ml_kem.h>
+#include "ml_kem_local.h"
+#ifndef NDEBUG
+# include <stdio.h>
+#endif
+
+/*
+ * Variant-independence macros, variously shortened for convenience and
+ * internal consumption by original BoringSSL code.
+ */
+# if BITS == 512
+# define kemsym(a, b)       a##_512_##b
+# elif BITS == 768
+# define kemsym(a, b)       a##_768_##b
+# elif BITS == 1024
+# define kemsym(a, b)       a##_1024_##b
+# else
+#  error "Unsupported ML-KEM variant bit size"
+# endif
+
+/* Constants */
+# define DEGREE         ML_KEM_DEGREE
+# define RANK           kemsym(ML_KEM,RANK)
+# define ETA1           kemsym(ML_KEM,ETA1)
+# define ETA2           kemsym(ML_KEM,ETA2)
+# define DU             kemsym(ML_KEM,DU)
+# define DV             kemsym(ML_KEM,DV)
+# define SECBITS        kemsym(ML_KEM,RNGSEC)
+# define CTEXT_BYTES    ML_KEM_CTEXT_BYTES(RANK, DU, DV)
+# define PUBKEY_BYTES   ML_KEM_PUBKEY_BYTES(RANK)
+# define PRVKEY_BYTES   ML_KEM_PRVKEY_BYTES(RANK)
+# define VECTOR_BYTES   ML_KEM_VECTOR_BYTES(RANK)
+
+# if (DEGREE != 256) || (ML_KEM_PRIME != 3329) || (ML_KEM_LOG2PRIME != 12) || \
+     (DU < 2) || (DU >= 12) || (DV < 2) || (DV >= 12) || \
+     (ETA1 < 2) || (ETA1 > 3) || (ETA2 < 2) || (ETA2 > 3)
+#  error "Invalid ML-KEM variant-independent constants"
+# endif
+# define U_VECTOR_BYTES (DU * RANK * DEGREE / 8)
+
+# if ETA1 == 2
+#  define cbd_eta1      ossl_ml_kem_cbd_2
+# elif ETA1 == 3
+#  define cbd_eta1      ossl_ml_kem_cbd_3
+# endif
+# if ETA2 == 2
+#  define cbd_eta2      ossl_ml_kem_cbd_2
+# elif ETA2 == 3
+#  define cbd_eta2      ossl_ml_kem_cbd_3
+# endif
+
+/* Types */
+typedef kemsym(ossl_ml_kem,vector)      vector;
+typedef kemsym(ossl_ml_kem,public_key)  public_key;
+typedef kemsym(ossl_ml_kem,private_key) private_key;
+
+/* Main keypair gen/encap/decap API */
+# define genkey_rand                kemsym(ossl_ml_kem,genkey_rand)
+# define encap_rand                 kemsym(ossl_ml_kem,encap_rand)
+# define decap                      kemsym(ossl_ml_kem,decap)
+
+/* For testing and encoding rather than KEM computation */
+# define genkey_seed                kemsym(ossl_ml_kem,genkey_seed)
+# define encap_seed                 kemsym(ossl_ml_kem,encap_seed)
+# define parse_public_key           kemsym(ossl_ml_kem,parse_public_key)
+# define encode_public_key          kemsym(ossl_ml_kem,encode_public_key)
+# define parse_private_key          kemsym(ossl_ml_kem,parse_private_key)
+# define encode_private_key         kemsym(ossl_ml_kem,encode_private_key)
+# define get_vinfo                  kemsym(ossl_ml_kem,get_vinfo)
+
+typedef const ossl_ml_kem_vinfo *vinfo_t;
```

Comment:
Suggest that type and function names are customised by the algorithm being implemented.  Over in #25990, I'm doing this with a `NAME` macro.  For example:

```c
#define NAME_CREATE3(n1, n2, n3, name)      NAME_CONCAT3(n1, n2, n3, name)
#define NAME(name)          NAME_CREATE3(NAME_1, NAME_2, PREFIX, name)

static void *NAME(newctx)(void *provctx)
```

Here it would just be including the BITS define as part of the name, likely using your `ossl_ml_kem_name` macro or similar.

Reply:
DEFINITELY NO. The purpose of the "vinfo" table is to make it possible for the algorithm implementation to be **identical** across the three algorithms, and just use the right slot in the parameter table. See https://github.com/openssl/openssl/pull/26172 for a continuation of that approach that significantly simplifies the code.

Reply:
The entire set of non-static libcrypto symbols for ML-KEM (all three variants) is now:
```
nm libcrypto.a | grep -w 'T'  | grep ml_kem | awk '{print $NF}' | sort -u
ossl_ml_kem_decap
ossl_ml_kem_encap_rand
ossl_ml_kem_encap_seed
ossl_ml_kem_encode_private_key
ossl_ml_kem_encode_public_key
ossl_ml_kem_genkey_rand
ossl_ml_kem_genkey_seed
ossl_ml_kem_get_vinfo
ossl_ml_kem_key_dup
ossl_ml_kem_key_free
ossl_ml_kem_key_new
ossl_ml_kem_parse_private_key
ossl_ml_kem_parse_public_key
ossl_ml_kem_pubkey_cmp
```